vms/proposervm/block/block.go

```diff
@@ -48,69 +69,168 @@ type statelessUnsignedBlock struct {
+	switch sb := any(b.StatelessBlock).(type) {
+	case statelessUnsignedBlock:
+		return sb.Certificate
+	case statelessUnsignedBlockV0:
+		return sb.Certificate
+	default:
+		panic(errUnexpectedStatelessUnsignedBlock)
+	}
 }
 
-func (b *statelessBlock) Bytes() []byte {
+func (b *statelessBlock[T]) innerTimestamp() int64 {
+	switch sb := any(b.StatelessBlock).(type) {
+	case statelessUnsignedBlock:
+		return sb.Timestamp
+	case statelessUnsignedBlockV0:
+		return sb.Timestamp
+	default:
+		panic(errUnexpectedStatelessUnsignedBlock)
+	}
+}
+
+func (b *statelessBlock[T]) Timestamp() time.Time {
+	return b.timestamp
+}
+
+func (b *statelessBlock[T]) Proposer() ids.NodeID {
+	return b.proposer
+}
+
+func (b *statelessBlock[T]) Bytes() []byte {
 	return b.bytes
 }
 
-func (b *statelessBlock) initialize(bytes []byte) error {
-	b.bytes = bytes
+func (b *statelessBlock[T]) ID() ids.ID {
+	return b.id
+}
 
-	// The serialized form of the block is the unsignedBytes followed by the
-	// signature, which is prefixed by a uint32. So, we need to strip off the
-	// signature as well as it's length prefix to get the unsigned bytes.
-	lenUnsignedBytes := len(bytes) - wrappers.IntLen - len(b.Signature)
-	unsignedBytes := bytes[:lenUnsignedBytes]
-	b.id = hashing.ComputeHash256Array(unsignedBytes)
+func (b *statelessBlock[T]) VerifySignature(pk *bls.PublicKey, parentVRFSig []byte, chainID ids.ID, networkID uint32) bool {
+	if pk == nil {
+		// proposer doesn't have a BLS key.
+		if len(parentVRFSig) == 0 {
+			// parent block had no VRF Signature.
+			// in this case, we verify that the current signature is empty.
+			return len(b.VRFSig()) == 0
+		}
+		// parent block had VRF Signature.
+		expectedSignature := NextHashBlockSignature(parentVRFSig)
+		return bytes.Equal(expectedSignature, b.VRFSig())
+	}
 
-	b.timestamp = time.Unix(b.StatelessBlock.Timestamp, 0)
-	if len(b.StatelessBlock.Certificate) == 0 {
-		return nil
+	// proposer does have a BLS key.
+	if len(parentVRFSig) == 0 {
+		// parent block had no VRF Signature.
+		msgHash := calculateBootstrappingBlockSig(chainID, networkID)
+		return bls.Verify(pk, b.vrfSig, msgHash[:])
 	}
+	return bls.Verify(pk, b.vrfSig, parentVRFSig)
+}
 
+func (b *statelessBlock[T]) initialize(bytes []byte) error {
 	var err error
-	b.cert, err = staking.ParseCertificate(b.StatelessBlock.Certificate)
+
+	b.bytes = bytes
+	if b.id, err = initializeID(bytes, b.Signature); err != nil {
+		return err
+	}
+
+	b.timestamp = time.Unix(b.innerTimestamp(), 0)
+	if len(b.innerCertificate()) == 0 {
+		return nil
+	}
+
+	b.cert, err = staking.ParseCertificate(b.innerCertificate())
 	if err != nil {
 		return fmt.Errorf("%w: %w", errInvalidCertificate, err)
 	}
 
 	b.proposer = ids.NodeIDFromCert(b.cert)
+
+	if vrfBytes := b.VRFSig(); len(vrfBytes) > 0 {
+		b.vrfSig, err = bls.SignatureFromBytes(vrfBytes)
+		if err != nil {
+			return fmt.Errorf("%w: %w", errFailedToParseVRFSignature, err)
```

Comment:
`b.vrfSig` may have been corrupted when the error occurred. Perhaps I'm being pedantic but I think it's worth setting it to `nil` in the error branch.

Reply:
I don't think that you're pedantic; I looked into that myself and confirmed that is already being guaranteed by `SignatureFromBytes`.
I really like when struct function like this are atomic ( i.e. function doesn't change any of the struct field in case of an error). In this case, however, I don't believe it would have a notable value.
If you'd like to see this as an atomic function, please let me know and I will update it accordingly.